### DISCUSSION TITLE
SocketIO 400 Bad Request messages handling

### DISCUSSION THREAD

**Discussion by amougnom** [miguelgrinberg/Flask-SocketIO#1811] [Q&A] [answered]
I use SocketIO_client ( 0.7.2) with Python3 to connect to Flask-SocketIO 4.3.1 server.

When I run python client manually sometimes, I got warning messages from server :

"WARNING:socketIO-client:xxxxxxxxxx/socket.io [connection error] unexpected status code (400 Bad Request)." And then emit doesn't work anymore....

When I run the same python program in background or as service the 400 Bad request is permanent and client can't receive message form server anymore.

I noticed that when I install eventlet on client, the client hangs on connect and no can't connect to server.

I need help to :

Fix theses messages at server sides
Handle warning messages at client side (read from code Warning messages and process them
Thanks.

**Comment by miguelgrinberg** [comment, marked as answer]
First of all, there is no need to spam all my projects with the question. One discussion is enough. Thank you.

Second, I don't know if the old socketio_client package works well. This is a very old/outdated client. I recommend that you use the client that comes with python-socketio, which you already have installed since that package is a dependency of Flask-SocketIO. If you use the same version of python-socketio for the server and the client you should be fine, I think.

**Reply by amougnom** [reply]
Hello,

Really sorry,  I created the first one before finding this which is more indicated for the topic. Unfortunately, I didn't find the option to delete the first one.

Thanks. I will update packages, test and update the ticket.

Regards,

**Comment by amougnom** [comment]
Hello,

 I installed new Falsk-Socketio (flask-socketio-5.1.1 python-socketio-5.5.2) but when trying to run the server  have following errors:

10.2.3.144 - - [05/Apr/2022 19:18:17] "GET /socket.io/?transport=polling&EIO=4&t=1649186297.098767 HTTP/1.1" 200 -
10.2.3.144 - - [05/Apr/2022 19:18:17] "GET /socket.io/?transport=websocket&EIO=4&sid=MvW9M4TdTDa7AeR8AAAA&t=1649186297.1027772 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/flask_socketio/__init__.py", line 43, in __call__
    return super(_SocketIOMiddleware, self).__call__(environ,
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/engineio/middleware.py", line 63, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/socketio/server.py", line 597, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/engineio/server.py", line 411, in handle_request
    packets = socket.handle_get_request(
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/engineio/socket.py", line 103, in handle_get_request
    return getattr(self, '_upgrade_' + transport)(environ,
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/engineio/socket.py", line 158, in _upgrade_websocket
    return ws(environ, start_response)
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/engineio/async_drivers/threading.py", line 22, in __call__
    self.ws = Server(environ)
  File "/python/dev/rest-api/rest-api/lib/python3.9/site-packages/simple_websocket/ws.py", line 268, in __init__
    raise RuntimeError('Cannot obtain socket from WSGI environment.')
RuntimeError: Cannot obtain socket from WSGI environment.

Please, could check the message and advise.

Thanks.

**Reply by amougnom** [reply]
Fixed by upgrading flask packages...

Now  need help because when customized  namespace don't work.  Only  "/" respond ...

**Reply by amougnom** [reply]
Hi,

Fixed custom namespace issue as well.

But I think  there is issue with new versions....

1) Server emits are not received by client unless client was connected and then restart server
2) Client disconnect by "CTRL+C" events are not processed by server like disconnect

I thought it was my program...So copied the example :  https://github.com/miguelgrinberg/Flask-SocketIO/blob/main/example/app_namespace.py

 I have exactly the same issue with this project app exple module. 

Versions : 

Server  : Flask-SocketIO     5.1.1,  python-socketio    5.5.2

Client:  python-socketio                   5.5.2,  websocket-client                  1.3.2

Thanks for helping.

**Reply by miguelgrinberg** [reply]
I see client disconnects reported in app.py and app_namespace.py. What is your client code?

**Comment by amougnom** [comment]
Hello,

Sorry for the silence. I was involved in an emergency.

I found what was the problem :

When a new client connect to the customized Namespace, the "on_connect" emit message ( I added an "emit" message to on_connect() of  customized Namespace class on server...) is not received by new connected client but all other clients connected to the same Namespace receive that  message. It's for that reason I thought the connection to the customized namespace failed but it succeed and others events and emit  are received by client.

If you could advise on why the emit of "on_connect"  is not received by new connected client that could help... But it's no more my main issue.

Thanks.

**Reply by miguelgrinberg** [reply]
The connect event is for the application to accept or reject the connection. When this event fires the client is not properly connected yet.

**Reply by amougnom** [reply]
Thanks for that information. I didn't know it. So this not an issue.

**Comment by amougnom** [comment]
Hi,

My new and urgent issue is:

1) Connection to Server failed several times with "One or more namespaces failed to connect" messages before connection succeed
   During that time I have many "HTTP/1.1" 400" errors message at server side

2) Client lost connection (is disconnected from server) several times ...

3) "HTTP/1.1 400" Errors messages at server side  :
 
"-" - - [22/Apr/2022:10:36:38 +0000] "POST /socket.io/?transport=polling&EIO=4&sid=StfVCRZ3BNgTCwn5AAAd HTTP/1.1" 400 17 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:37:39 +0000] "GET /socket.io/?transport=polling&EIO=4&t=1650623562.2356913 HTTP/1.1" 200 97 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:37:40 +0000] "GET /socket.io/?transport=websocket&EIO=4&sid=1kr453jkDFSbM7uwAAAX&t=1650623562.4438465 HTTP/1.1" 400 17 "-" "-"
"-" - - [22/Apr/2022:10:37:40 +0000] "POST /socket.io/?transport=polling&EIO=4&sid=1kr453jkDFSbM7uwAAAX HTTP/1.1" 400 17 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:38:04 +0000] "GET /socket.io/?transport=polling&EIO=4&sid=1kr453jkDFSbM7uwAAAX&t=1650623563.176217 HTTP/1.1" 200 1 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:38:41 +0000] "GET /socket.io/?transport=polling&EIO=4&t=1650623624.179965 HTTP/1.1" 200 97 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:38:42 +0000] "GET /socket.io/?transport=websocket&EIO=4&sid=8mj5cnAmefLd45hvAAAY&t=1650623624.3888023 HTTP/1.1" 400 17 "-" "-"
"-" - - [22/Apr/2022:10:38:42 +0000] "GET /socket.io/?transport=polling&EIO=4&sid=8mj5cnAmefLd45hvAAAY&t=1650623625.1719778 HTTP/1.1" 400 17 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:38:42 +0000] "POST /socket.io/?transport=polling&EIO=4&sid=8mj5cnAmefLd45hvAAAY HTTP/1.1" 400 17 "-" "python-requests/2.25.1"
"-" - - [22/Apr/2022:10:39:43 +0000] "GET /socket.io/?transport=polling&EIO=4&t=1650623686.1759317 HTTP/1.1" 200 97 "-" "python-requests/2.25.1"

I have same message even when connection to server IP or through NGINX Proxy. 

I implemented server with gunicorn==20.1.0 and eventlet==0.30.2.

**Server code :** 

from flask import Flask
from flask_socketio import SocketIO,send,emit,disconnect, Namespace

app = Flask(__name__)

class MyNamespaceClass(Namespace):

	def on_connect(self):
		self.emit('connected')
		
	def on_test_emit(self):
	  print('Client sent a message')

socketio = SocketIO(app,logger=True, engineio_logger=True,cors_allowed_origins="*",async_mode='eventlet', ping_timeout=60)

socketio.on_namespace(MyNamespaceClass('/Test'))

if __name__ == '__main__':
	socketio.run(app,debug=True,host='0.0.0.0')

**Client Code:**

import socketio
host =  [IP:port]

class PMonitor(socketio.ClientNamespace):

    def on_connect(self):
        print('Connected to server')

sio = socketio.Client()
sio.connect(host, namespaces=['/Test'],wait_timeout = 60)
sio.register_namespace(PMonitor('/Test'))
sio.emit('test_emit',namespace='/Test')
sio.wait()

Thanks  for your availability and patience.

Regards.

**Reply by miguelgrinberg** [reply]
The server has a `/Testing` namespace. The client is connecting to `/Test`.

**Reply by amougnom** [reply]
Sorry,

This a long code for production, I  just copied and and  changed names for public message. In real code this the same name and connection  succeed after some attempts.

If the name where different  in production , I would never be able to connect.

Thanks.

**Comment by amougnom** [comment]
Also, more details:  I doesn't have this problem when testing apps in local (by running apps by pyhton wsgi.py).

The problem occur only when  deployed on remote server with Gunicorn and gevent.

Thanks.

**Reply by miguelgrinberg** [reply]
Sorry but I cannot help, since you aren't providing enough information. If you have issues with deployment, please consult the deployment section of the documentation, which shows in detail what commands you can use.

**Reply by amougnom** [reply]
> Sorry but I cannot help, since you aren't providing enough information. If you have issues with deployment, please consult the deployment section of the documentation, which shows in detail what commands you can use.

Please,  tell  me which information you need, I will provide them all easily .

I thought I provided all required information, versions , deployment tools , server and client codes, servers error logs....

Just tell  me what is required for support.

Thanks.

**Reply by miguelgrinberg** [reply]
Please read the documentation. The instructions are all written there.

**Comment by amougnom** [comment]
Hello,

You are right!

I checked again and my bad was to trying to use multiple workers on single node. I changed workers=1 in gunicorn config file and it fixed all  issues.

You can clause this ticket right now please.

Thanks.
